src/sprite_resource.c

```diff
@@ -0,0 +1,884 @@
+{
+    MultiSpriteResourceData *multiSpriteData = Heap_AllocFromHeap(heapID, sizeof(MultiSpriteResourceData));
+    NNS_G2dGetUnpackedMultiCellBank(rawData, &multiSpriteData->multiSpriteBank);
+
+    return multiSpriteData;
+}
+
+static MultiSpriteAnimResourceData *SpriteUtil_UnpackMultiSpriteAnimResource(void *rawData, enum HeapId heapID)
+{
+    MultiSpriteAnimResourceData *multiSpriteAnimData = Heap_AllocFromHeap(heapID, sizeof(MultiSpriteAnimResourceData));
+    NNS_G2dGetUnpackedMCAnimBank(rawData, &multiSpriteAnimData->multiSpriteAnimBank);
+
+    return multiSpriteAnimData;
+}
+
+static void SpriteResource_FreeData(SpriteResource *spriteRes)
+{
+    if (spriteRes->data) {
+        Heap_FreeToHeap(spriteRes->data);
+    }
+
+    spriteRes->data = NULL;
+}
+
+static void *SpriteResource_GetData(const SpriteResource *spriteRes)
+{
+    GF_ASSERT(spriteRes);
+    GF_ASSERT(spriteRes->data);
+
+    return spriteRes->data;
+}
+
+static void SpriteResourceCollection_InitResFromFile(SpriteResourceCollection *spriteResources, SpriteResource *spriteRes, const char *filename, int id, NNS_G2D_VRAM_TYPE vramType, int paletteIdx, enum SpriteResourceType type, enum HeapId heapID)
+{
+    spriteRes->rawResource = ResourceCollection_AddFromFile(spriteResources->collection, filename, id, heapID);
+    spriteRes->type = type;
+
+    SpriteResource_UnpackData(spriteRes, type, vramType, paletteIdx, heapID);
+}
+
+static void SpriteResourceCollection_InitRes(SpriteResourceCollection *spriteResources, SpriteResource *spriteRes, int narcIdx, int memberIdx, BOOL compressed, int id, NNS_G2D_VRAM_TYPE vramType, int paletteIdx, enum SpriteResourceType type, enum HeapId heapID, u32 param10)
+{
+    void *data = sub_02006FE8(narcIdx, memberIdx, compressed, heapID, param10);
+
+    spriteRes->rawResource = ResourceCollection_Add(spriteResources->collection, data, id);
+    spriteRes->type = type;
+
+    SpriteResource_UnpackData(spriteRes, type, vramType, paletteIdx, heapID);
+}
+
+static void SpriteResourceCollection_InitResFromNARC(SpriteResourceCollection *spriteResources, SpriteResource *spriteRes, NARC *narc, int memberIdx, BOOL compressed, int id, NNS_G2D_VRAM_TYPE vramType, int paletteIdx, enum SpriteResourceType type, enum HeapId heapID, u32 param10)
+{
+    void *data = SpriteUtil_ReadNARCMember(narc, memberIdx, compressed, heapID, param10);
+
+    spriteRes->rawResource = ResourceCollection_Add(spriteResources->collection, data, id);
+    spriteRes->type = type;
+
+    SpriteResource_UnpackData(spriteRes, type, vramType, paletteIdx, heapID);
+}
+
+static int SpriteResourceTableEntryNARC_GetEntryCount(const SpriteResourceTableEntryNARC *entries)
+{
+    int i = 0;
+    while (entries[i].narcIndex != NARC_INDEX_NONE) {
+        i++;
+    }
+
+    return i;
+}
+
+static void *SpriteUtil_ReadNARCMember(NARC *narc, u32 memberIdx, BOOL compressed, u32 heapID, u32 param4)
```

Comment:
**suggestion:** Name `param4` (and the associated params in other functions); seems to me that it's a bool that designates whether or not an allocation should happen at the end of the heap or not.